packages/frontend/src/state/lp/hooks.ts

```diff
@@ -177,6 +178,159 @@ export const useCollectFees = () => {
   return collectFees
 }
 
+// Rebalance via general swap
+export const useRebalanceGeneralSwap = () => {
+  const address = useAtomValue(addressAtom)
+  const controllerHelperContract = useAtomValue(controllerHelperHelperContractAtom)
+  const { controllerHelper, weth, oSqueeth, squeethPool } = useAtomValue(addressesAtom)
+  const controllerContract = useAtomValue(controllerContractAtom)
+  const handleTransaction = useHandleTransaction()
+  const isWethToken0 = useAtomValue(isWethToken0Atom)
+  const getDebtAmount = useGetDebtAmount()
+  const getVault = useGetVault()
+  const getDecreaseLiquidity = useGetDecreaseLiquidity()
+  const getPosition = useGetPosition()
+  const getTwapSqueethPrice = useGetTwapSqueethPrice()
+  const squeethPoolContract = useAtomValue(squeethPoolContractAtom)
+  const getQuote = useGetQuote()
+  const rebalanceGeneralSwap = useAppCallback(
+    async (vaultId: number, lowerTickInput: number, upperTickInput: number, onTxConfirmed?: () => void) => {
+      const vaultBefore = await getVault(vaultId)
+      const uniTokenId = vaultBefore?.NFTCollateralId 
+      const position = uniTokenId ? await getPosition(uniTokenId) : null
+      if (!controllerContract || !controllerHelperContract || !address || !position || !vaultBefore || !squeethPoolContract) return
+      const shortAmount = fromTokenAmount(vaultBefore.shortAmount, OSQUEETH_DECIMALS)
+      const debtInEth = await getDebtAmount(shortAmount)
+      const collateralToFlashloan = debtInEth.multipliedBy(COLLAT_RATIO)
+
+      const amount0Min = new BigNumber(0)
+      const amount1Min = new BigNumber(0)
+
+      const lowerTick = nearestUsableTick(lowerTickInput, TICK_SPACE)
+      const upperTick = nearestUsableTick(upperTickInput, TICK_SPACE)
+
+      // Get current LP positions
+      const { amount0, amount1 } = await getDecreaseLiquidity(uniTokenId, position.liquidity, 0, 0, Math.floor(Date.now() / 1000 + 86400))
```

Comment:
we can also calculate these directly like dis:
if p < pa:
    amount0, amount1 = L*(1/np.sqrt(pa)- 1/np.sqrt(pb)), 0)
elif p < pb:
    amount0, amount1 = L*(1/np.sqrt(p)- 1/np.sqrt(pb)), L*(np.sqrt(p)- np.sqrt(pa))
else:
    amount0, amount1 = 0, L*(np.sqrt(pb)- np.sqrt(pa))